### DISCUSSION TITLE
How to achieve a good compression ratio when saving to zarr

### DISCUSSION THREAD

**Discussion by marstonsward** [pydata/xarray#7570] [Q&A] [answered]
Curious how `Blosc` compression works when appending to a `DirectoryStore` using `to_zarr()`. An example pseudocode
```python
import xarray as xr
import zarr as za

nlat = 700
nlon = 2000
nlev =20
chunks3d={'time':1, 'feature': 1, 'lev': nlev, 'lat': nlat//7, 'lon': nlon//10}
chunks2d={'time':1, 'feature': 1, 'lev': 1, 'lat': nlat//7, 'lon': nlon//10}
compressor = za.Blosc(cname="zstd", clevel=9, shuffle=2)
g3d = 'f3d'
g2d = 'f2d'
store = 'test.zarr'

Append = False
# Each ts in datasets represent a unique variable but is generically called 'varb3d'
# Each varb3d dims: (time=1, feature=1, nlev=nlev, lat=nlat, lon=nlon)
for ts in datasets:
    if Append:
         ts.chunk(chunks3d).to_zarr(store, group=g3d, append_dim='feature', consolidated=True)
    else:
         ts.chunk(chunks3d).to_zarr(store, group=g3d, consolidated=True, encoding={'varb3d': {"compressor": compressor}})
         Append = True
```
The above code works but the compression level change from 6 to 9 didn't reduce the size by much. So, I'm wondering if the first variable in the loop is actually compressed? It is not intuitive that it is. If I add the compression in initial creation phase of the file, I get a key error. Curious if the whole array is compressed during the append process or will only appended arrays are compressed?
I'm guessing there is some copying going on under the hood and all the variables are compressed in the end.

**Comment by rabernat** [comment, marked as answer]
I don't know what your data source is, but in general, floating point data don't compress well using lossless compression. I would not expect any meaningful size difference between clevel 6 vs 9.

**Reply by marstonsward** [reply]
@rabernat 

Thank you for the reply. Yes, my data is all float32. What compression choice would you suggest?

**Reply by rabernat** [reply]
If you want to achieve any significant compression ratios, you will need to accept lossy compression. [Xbitinfo](https://github.com/observingClouds/xbitinfo) is an attractive option right now. zfp is also good.

**Reply by marstonsward** [reply]
@rabernat 
Thank you. I did some research on this and read the article from LLNL. I would like to test lossy. zfpy is in zarr but there is no documentation on its usage. Is this the same as zfp? Xbitinfo doesn't seem to be available in zarr. I write to zarr via the Xarray `to_zarr()`.

**Reply by rabernat** [reply]
Yes, xbitinfo will work fine with Zarr. Xbitinfo just rounds the data to the specified bit precision. Then you can pass it off to Zarr's standard lossless compressors.

**Reply by marstonsward** [reply]
This seems straightforward enough, but how is this read back into an uncompressed dataArray or dataset?

```
import xarray as xr
import xbitinfo as xb
ds = xr.tutorial.load_dataset(inpath)
bitinfo = xb.get_bitinformation(ds, dim="lon")  # calling bitinformation.jl.bitinformation
keepbits = xb.get_keepbits(bitinfo, inflevel=0.99)  # get number of mantissa bits to keep for 99% real information
ds_bitrounded = xb.xr_bitround(ds, keepbits)  # bitrounding keeping only keepbits mantissa bits
ds_bitrounded.to_compressed_netcdf(outpath)  # save to netcdf with compression
```

**Reply by rabernat** [reply]
> This seems straightforward enough, but how is this read back into an uncompressed dataArray or dataset?

Reading is super simple. You just open the file and read it. No special decompression is required.

**Reply by marstonsward** [reply]
Super!
Seems almost too good to be true. Will test this and see if it is as advertised.
Thank you.

**Reply by marstonsward** [reply]
One last question on this part:
```
bitinfo = xb.get_bitinformation(ds, dim="lon")  # calling bitinformation.jl.bitinformation
```
Is the compression applied here via a specified dim? So if I have an array with dims: `time,xx,lev,lat,lon`, how is this handled? All the data is covered by the `lat and lon` but what about the other dims?

**Reply by rabernat** [reply]
You've exceeded my knowledge. Sounds like a good question for the xbitinfo issue tracker! 😉

**Reply by marstonsward** [reply]
Hehe...indeed. Will post there. Thanks for your tremendous help hitherto :-)
